File: source/includes/write/update.cpp
#include <iostream>

#include <bsoncxx/builder/basic/document.hpp>
#include <bsoncxx/json.hpp>
#include <mongocxx/client.hpp>
#include <mongocxx/instance.hpp>
#include <mongocxx/uri.hpp>

using bsoncxx::builder::basic::kvp;
using bsoncxx::builder::basic::make_document;

int main() {
    mongocxx::instance instance;
    mongocxx::uri uri("<connection string>");
    mongocxx::client client(uri);

    // start-db-coll
    auto db = client["sample_restaurants"];
    auto collection = db["restaurants"];
    // end-db-coll

    {
        // Updates the "name" value of a document from "Bagels N Buns" to "2 Bagels 2 Buns"
        // start-update-one
        auto query_filter = make_document(kvp("name", "Bagels N Buns"));
        auto update_doc = make_document(kvp("$set", make_document(kvp("name", "2 Bagels 2 Buns"))));

        auto result = collection.update_one(query_filter.view(), update_doc.view());
        // end-update-one
    }

    {
        // Updates the "cuisine" value of documents from "Pizza" to "Pasta"
        // start-update-many
        auto query_filter = make_document(kvp("cuisine", "Pizza"));
        auto update_doc = make_document(kvp("$set", make_document(kvp("cuisine", "Pasta"))));

        auto result = collection.update_many(query_filter.view(), update_doc.view());
        // end-update-many
    }

    {

        // Updates the "borough" value of matching documents and inserts a document if none match
        // start-update-options
        mongocxx::options::update opts{};
        opts.upsert(true);

        auto query_filter = make_document(kvp("borough", "Manhattan"));
        auto update_doc = make_document(kvp("$set", make_document(kvp("borough", "Manhattan (north)"))));

        auto result = collection.update_many(query_filter.view(), update_doc.view(), opts);
        // end-update-options
    }

    {

        // Updates the "name" value of matching documents and prints the number of updates
        // start-update-result
        auto query_filter = make_document(kvp("name", "Dunkin' Donuts"));
        auto update_doc = make_document(kvp("$set", make_document(kvp("name", "Dunkin'"))));

        auto result = collection.update_many(query_filter.view(), update_doc.view());
        std::cout << "Modified documents: " << result->modified_count() << std::endl;
        // end-update-result
    }
}

File: source/includes/write/update.scala
import org.mongodb.scala._
import org.mongodb.scala.bson.Document
import org.mongodb.scala.model.Filters.equal
import org.mongodb.scala.model.UpdateOptions
import org.mongodb.scala.model.Updates.{combine, rename, set}
import org.mongodb.scala.result.UpdateResult

object Update {

  def main(args: Array[String]): Unit = {
    val mongoClient = MongoClient("<connection string URI>")

    // start-db-coll
    val database: MongoDatabase = mongoClient.getDatabase("sample_restaurants")
    val collection: MongoCollection[Document] = database.getCollection("restaurants")
    // end-db-coll

    // start-update-one
    val filter = equal("name", "Happy Garden")
    val update = set("name", "Mountain House")
    val observable: Observable[UpdateResult] = collection.updateOne(filter, update)

    observable.subscribe(new Observer[UpdateResult] {
      override def onNext(result: UpdateResult): Unit =
          println(s"Updated document count: ${result.getModifiedCount}")
      override def onError(e: Throwable): Unit = println(s"Failed: ${e.getMessage}")
      override def onComplete(): Unit = println("Completed")
    })
    // end-update-one

    // start-update-many
    val filter = equal("name", "Starbucks")
    val update = rename("address", "location")
    val observable: Observable[UpdateResult] = collection.updateMany(filter, update)

    observable.subscribe(new Observer[UpdateResult] {
      override def onNext(result: UpdateResult): Unit =
          println(s"Updated document count: ${result.getModifiedCount}")
      override def onError(e: Throwable): Unit = println(s"Failed: ${e.getMessage}")
      override def onComplete(): Unit = println("Completed")
    })
    // end-update-many

    // start-update-options
    val filter = equal("name", "Sunrise Pizzeria")
    val opts = UpdateOptions().upsert(true)
    val update = combine(
      set("borough", "Queens"),
      set("cuisine", "Italian")
    )
    val observable: Observable[UpdateResult] = collection.updateOne(filter, update, opts)

    observable.subscribe(new Observer[UpdateResult] {
      override def onNext(result: UpdateResult): Unit =
          println(s"Updated document count: ${result.getModifiedCount}")
      override def onError(e: Throwable): Unit = println(s"Failed: ${e.getMessage}")
      override def onComplete(): Unit = println("Completed")
    })
    // end-update-options

    Thread.sleep(1000)
    mongoClient.close()
  }
}
